### PR TITLE
feature: support qwen2.5 fuction_call

### PR DESCRIPTION
**Pull request by akai-shuuichi**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
https://github.com/InternLM/lmdeploy/issues/2610 的后续，支持了Qwen2.5的fuctioncall，参考了 https://github.com/InternLM/lmdeploy/pull/2147 和Qwen2.5的chatTemplate [tokenizer_config.json](https://www.modelscope.cn/models/Qwen/Qwen2.5-14B-Instruct/file/view/master?fileName=tokenizer_config.json&status=1)

## Modification

        modified:   lmdeploy/model.py
        modified:   lmdeploy/serve/async_engine.py
        modified:   lmdeploy/serve/openai/api_server.py

## Use cases (Optional)
基于lmdeployv0.6.2.post1

```
lmdeploy serve api_server  Qwen2.5-14B-Instruct-AWQ --tp 2
```

不携带tools参数
``` python
from openai import OpenAI
client = OpenAI(
    api_key='YOUR_API_KEY',
    base_url="http://0.0.0.0:23333/v1"
)
model_name = client.models.list().data[0].id
response = client.chat.completions.create(
  model=model_name,
  messages=[
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": " provide three suggestions about time management"},
  ],
    temperature=0.8,
    top_p=0.8
)
print(response)

```
得到类似的结果：
``` python
ChatCompletion(id='2', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content="Certainly! Effective time management can greatly enhance productivity and reduce stress. Here are three suggestions for better time management:\n\n1. **Prioritize Tasks with a To-Do List**: Start your day by listing all the tasks you need to accomplish. Then prioritize these tasks based on their importance and urgency. This helps you focus on what truly matters and prevents you from getting overwhelmed by a long list of tasks. Tools like digital planners or simple pen-and-paper lists can be very effective.\n\n2. **Implement the Pomodoro Technique**: This technique involves working for a set period (typically 25 minutes), followed by a short break (5 minutes). After four such cycles, take a longer break (15-30 minutes). This method can help maintain high levels of concentration and prevent burnout. It's particularly useful for tasks that require a lot of mental effort.\n\n3. **Use Time Blocking**: Allocate specific blocks of time for different activities throughout your day. For example, you might dedicate the early morning hours to tasks that require the most concentration, mid-morning to meetings or calls, and later in the day to less demanding tasks or administrative work. This approach helps in managing time more efficiently and can also help in maintaining a better work-life balance.\n\nBy implementing these strategies, you can improve your time management skills and achieve more in less time.", refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None))], created=1731635562, model='Qwen2.5-14B-Instruct-AWQ', object='chat.completion', service_tier=None, system_fingerprint=None, usage=CompletionUsage(completion_tokens=271, prompt_tokens=35, total_tokens=306, completion_tokens_details=None, prompt_tokens_details=None))
```

实际拼接好的prompt(qwen2.5系列的sysprompt和之前的系列不一样)

```
<|im_start|>system
You are Qwen, created by Alibaba Cloud. You are a helpful assistant.<|im_end|>
<|im_start|>user
 provide three suggestions about time management<|im_end|>
<|im_start|>assistant
```

携带tools参数:

```python
from openai import OpenAI
import json

def get_current_temperature(arguments):
    # print(json.dumps(arguments))
    if json.loads(arguments).get('location') == 'San Francisco, CA, USA':
        return "26.0­°C"
    else:
        return "27.0"

def get_temperature_date(arguments):
    if json.loads(arguments)['location'] == 'San Francisco, CA, USA' and json.loads(arguments)['location'] == 'San Francisco, CA, USA':
        return "26.5­°C"
    else:
        return "27.5­°C"

tools = [{
    'type': 'function',
    'function': {
        'name': 'get_current_temperature',
        'description': 'Get current temperature at a location.',
        'parameters': {
            'type': 'object',
            'properties': {
                'location': {
                    'type': 'string',
                    'description': 'The location to get the temperature for, in the format \'City, State, Country\'.'
                },
                'unit': {
                    'type': 'string',
                    'enum': [
                        'celsius',
                        'fahrenheit'
                    ],
                    'description': 'The unit to return the temperature in. Defaults to \'celsius\'.'
                }
            },
            'required': [
                'location'
            ]
        }
    }
}, {
    'type': 'function',
    'function': {
        'name': 'get_temperature_date',
        'description': 'Get temperature at a location and date.',
        'parameters': {
            'type': 'object',
            'properties': {
                'location': {
                    'type': 'string',
                    'description': 'The location to get the temperature for, in the format \'City, State, Country\'.'
                },
                'date': {
                    'type': 'string',
                    'description': 'The date to get the temperature for, in the format \'Year-Month-Day\'.'
                },
                'unit': {
                    'type': 'string',
                    'enum': [
                        'celsius',
                        'fahrenheit'
                    ],
                    'description': 'The unit to return the temperature in. Defaults to \'celsius\'.'
                }
            },
            'required': [
                'location',
                'date'
            ]
        }
    }
}]
messages = [{'role': 'user', 'content': 'Today is 2024-11-14, What\'s the temperature in San Francisco now? How about tomorrow?'}]

client = OpenAI(api_key='YOUR_API_KEY', base_url='http://0.0.0.0:23333/v1')
model_name = client.models.list().data[0].id
response = client.chat.completions.create(
    model=model_name,
    messages=messages,
    temperature=0.8,
    top_p=0.8,
    stream=False,
    tools=tools)
print(response.choices[0].message.tool_calls)
messages.append(response.choices[0].message)

for tool_call in response.choices[0].message.tool_calls:
    if tool_call.function.name == 'get_temperature_date':
        content = get_temperature_date(tool_call.function.arguments)
    if tool_call.function.name == 'get_current_temperature':
        content = get_current_temperature(tool_call.function.arguments)
    messages.append({
            'role': 'tool',
            'name': tool_call.function.name,
            'content': content,
            'tool_call_id': tool_call.id
        })


response = client.chat.completions.create(
    model=model_name,
    messages=messages,
    temperature=0.8,
    top_p=0.8,
    stream=False,
    tools=tools)
print(response.choices[0].message.content)
```

result：
```

[ChatCompletionMessageToolCall(id='0', function=Function(arguments='{"location": "San Francisco, CA, USA"}', name='get_current_temperature'), type='function'),
ChatCompletionMessageToolCall(id='1', function=Function(arguments='{"location": "San Francisco, CA, USA", "date": "2024-11-15", "unit": "celsius"}', name='get_temperature_date'), type='function')]

The current temperature in San Francisco is 26.0°C. The temperature is predicted to be 26.5°C tomorrow, November 15, 2024. haha

```

最后一次调用的prompt：

```
<|im_start|>system
You are Qwen, created by Alibaba Cloud. You are a helpful assistant.

# Tools

You may call one or more functions to assist with the user query.

You are provided with function signatures within <tools></tools> XML tags:
<tools>
{"description": "Get current temperature at a location.", "name": "get_current_temperature", "parameters": {"type": "object", "properties": {"location": {"type": "string", "description": "The location to get the temperature for, in the format 'City, State, Country'."}, "unit": {"type": "string", "enum": ["celsius", "fahrenheit"], "description": "The unit to return the temperature in. Defaults to 'celsius'."}}, "required": ["location"]}}
{"description": "Get temperature at a location and date.", "name": "get_temperature_date", "parameters": {"type": "object", "properties": {"location": {"type": "string", "description": "The location to get the temperature for, in the format 'City, State, Country'."}, "date": {"type": "string", "description": "The date to get the temperature for, in the format 'Year-Month-Day'."}, "unit": {"type": "string", "enum": ["celsius", "fahrenheit"], "description": "The unit to return the temperature in. Defaults to 'celsius'."}}, "required": ["location", "date"]}}
</tools>

For each function call, return a json object with function name and arguments within <tool_call></tool_call> XML tags:
<tool_call>
{"name": <function-name>, "arguments": <args-json-object>}
</tool_call><|im_end|>
<|im_start|>user
Today is 2024-11-14, What's the temperature in San Francisco now? How about tomorrow?<|im_end|>
<|im_start|>assistant

<tool_call>
{"name": "get_current_temperature", "arguments": {"location": "San Francisco, CA, USA"}}
</tool_call>
<tool_call>
{"name": "get_temperature_date", "arguments": {"location": "San Francisco, CA, USA", "date": "2024-11-15", "unit": "celsius"}}
</tool_call><|im_end|>
<|im_start|>user
<tool_response>
26.0­°C
</tool_response>
<tool_response>
26.5­°C
</tool_response>
<|im_end|>
<|im_start|>assistant
```

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by akai-shuuichi** [comment]
尝试请求了本地，是可以正常运行的
`curl --location --request POST 'http://127.0.0.1:9000/v1/chat/completions' \
--header 'Connection: keep-alive' \
--header 'Content-Type: application/json' \
--data-raw '{
    "temperature": 0,
    "messages": [
        {
            "role": "user",
            "content": "告诉我武汉天气如何"
        }
    ],
    "model": "Qwen2.5-14B-Instruct-AWQ",
    "tools": [
        {
            "type": "function",
            "function": {
                "name": "get_current_weather",
                "description": "Get the current weather in a given location",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "location": {
                            "type": "string",
                            "description": "The city and state, e.g. San Francisco, CA"
                        }
                    },
                    "required": [
                        "location"
                    ]
                }
            }
        }
    ]
}'`
return
`{"id":"2","object":"chat.completion","created":1731310565,"model":"Qwen2.5-14B-Instruct-AWQ","choices":[{"index":0,"message":{"role":"assistant","content":"","tool_calls":[{"id":"0","type":"function","function":{"name":"get_current_weather","arguments":"{\"location\": \"\\u6b66\\u6c49, China\"}"}}]},"logprobs":null,"finish_reason":"tool_calls"}],"usage":{"prompt_tokens":184,"total_tokens":205,"completion_tokens":21}}`

**Comment by lvhan028** [comment]
Please resolve the linting issue
```shell
pip install pre-commit
cd lmdeploy  # the root dir of lmdeploy repo
pre-commit install
pre-commit run --all-files
```

**Comment by akai-shuuichi** [comment]
存在一些NPE, 解决中。

**Comment by akai-shuuichi** [comment]
找到个bug，只能解析一个tool，正在解决中

**Comment by akai-shuuichi** [comment]
目前的tools似乎只支持单个调用，我是否应该在pr修改支持同时调用多个函数？

File: lmdeploy/model.py [hunk 1]
```diff
@@ -892,6 +892,109 @@ def match(cls, model_path: str) -> Optional[str]:
...
+        return ret
+
+    @classmethod
+    def match(cls, model_path: str) -> Optional[str]:
+        """Return the model_name that was registered to MODELS.
+
+        Args:
+            model_path (str): the model path used for matching.
+        """
+        if 'qwen2.5' in model_path.lower():
```
Comment (AllentDan) [inline lmdeploy/model.py:994]: may add `qwen2_5` as well

File: lmdeploy/serve/async_engine.py [hunk 2]
```diff
@@ -652,6 +652,16 @@ def parse_tool_response(self, text, tools, **kwargs):
...
+        elif '<tool_call>' in text:  # qwen2.5
+            action = text.split('</tool_call>')[0]
+            action = action.split('<tool_call>')
+            if len(action) == 1:
+                text, action = '', action[0]
+            else:
+                text, action = action
+            action = json.loads(action)
+            name, parameters = action['name'], json.dumps(
+                action.get('parameters', action.get('arguments', {})))
```
Comment (AllentDan) [inline lmdeploy/serve/async_engine.py:664]: may add `ensure_ascii=False` to json.dumps

File: lmdeploy/model.py [hunk 3]
```diff
@@ -892,6 +892,109 @@ def match(cls, model_path: str) -> Optional[str]:
...
+                    or (message['role'] == 'system' and index != 0)
+                    or (message['role'] == 'assistant'
+                        and message.get('tools_call') is not None)):
+                ret += f"{box_map[message['role']]}{message['content']}{self.eoh}"
+            if message['role'] == 'assistant':
+                ret += f"{box_map[message['role']]}"
+                if message.get('content') is not None:
+                    ret += f"\n{message['content']}"
+            if message.get('tools_call') is not None:
+                toolsCall = message['tools_call']
```
Comment (AllentDan) [inline lmdeploy/model.py:972]: toolsCall -> tools_call and the same for the following variables.

**Comment by AllentDan** [comment]
不是每个模型都支持多函数，你可以为 qwen2.5 添加支持

File: lmdeploy/model.py [hunk 4]
```diff
@@ -927,14 +927,119 @@ def match(cls, model_path: str) -> Optional[str]:
...
+                    or (message['role'] == 'assistant'
+                        and message.get('tools_call') is not None)):
+                ret += f"{box_map[message['role']]}{message['content']}{self.eoh}"
+            if message['role'] == 'assistant':
+                ret += f"{box_map[message['role']]}"
+                if message.get('content') is not None:
+                    ret += f"\n{message['content']}"
+            if message.get('tools_call') is not None:
+                tools_call = message['tools_call']
+                for toolCall in tools_call:
```
Comment (AllentDan) [inline lmdeploy/model.py:1017]: tool_call

**Comment by akai-shuuichi** [comment]
> 不是每个模型都支持多函数，你可以为 qwen2.5 添加支持

谢谢回复！准备使用正则表达式去实现它
```
pattern = r"<tools>(.*?)</tools>"

tools = re.findall(pattern, text)
```

**Comment by akai-shuuichi** [comment]
参考https://github.com/QwenLM/Qwen2.5/discussions/1073


1. 预期内的返回
```
original text：
<tool_call>tool1_info<tool_call><tool_call>tool2_info<tool_call>
```

处理效果：
```
{
    "content": "",
    "tool_calls":[
        tool1_info,
        tool2_info
    ]
}
```


此外，为了处理一些在标记<tool_call>之外生成的数据，我为这几种情况做了兼容处理：
1. 多余文本出现在了最前面
```
original text：
text1<tool_call>tool1_info<tool_call><tool_call>tool2_info<tool_call>
```


2. 多余文本出现在了最后面
```
original text：
<tool_call>tool1_info<tool_call><tool_call>tool2_info<tool_call>text1
```

这两种情况的处理效果都是：
```
{
    "content": text1,
    "tool_calls":[
        tool1_info,
        tool2_info
    ]
}
```


需要注意的是，模型并不能100%保证生成结果符合预期，因此除了上面三种情况以外，还可能出现其他的问题，比如

```
original text：
text1<tool_call>tool1_info<tool_call>text2<tool_call>tool2_info<tool_call>text3
```

这种情况是没办法解决的，我只解析出来了text1


此外，我将parse_tool_response的返回结构改成了 str, List[tuple] 以供外部解析

File: lmdeploy/serve/async_engine.py [hunk 5]
```diff
@@ -648,25 +649,37 @@ def parse_tool_response(self, text, tools, **kwargs):
...
+            for match_result in match_result_list:
+                action = json.loads(match_result)
+                call_info_list.append((action['name'],
+                                       json.dumps(action['arguments'],
+                                                  ensure_ascii=False)))
+            # get text outside of tags
+            if not text.startswith('<tool_call>'):
+                text = text[:text.find('<tool_call>')]
+            elif not text.endswith('</tool_call>'):
+                text = text[text.find('</tool_call>') + len('</tool_call>'):]
```
Comment (AllentDan) [inline lmdeploy/serve/async_engine.py:672]: Does it extract real text when there are multiple tool_calls in the response?

File: lmdeploy/serve/async_engine.py [hunk 6]
```diff
@@ -648,25 +649,37 @@ def parse_tool_response(self, text, tools, **kwargs):
...
-            action = text.split('</tool_call>')[0]
-            action = action.split('<tool_call>')
-            if len(action) == 1:
-                text, action = '', action[0]
+            call_info_list = [(name, parameters)]
+        elif '<tool_call>' in text and '</tool_call>' in text:  # qwen2.5
+            # get tool_call in text
+            pattern = r'<tool_call>(.*?)</tool_call>'
+            match_result_list = re.findall(pattern, text, re.DOTALL)
+            call_info_list = []
```
Comment (AllentDan) [inline lmdeploy/serve/async_engine.py:662]: May update document for the multiple tool_calls response.

File: lmdeploy/serve/async_engine.py (hunk 5 above)
Comment (akai-shuuichi) [reply lmdeploy/serve/async_engine.py:672]: sorry, it should be rfind

**Comment by akai-shuuichi** [comment]
我忽略了一点，在多函数调用的场景下，模型可能无法区分出来多个tool的回复目标（即使指定了函数名称， 也会有名称冲突的风险），一种简单做法应该是将name和tool_call_id放到prompt中，tool_call_id应该是随机生成的id而不用在函数列表中获取下标位置

**Comment by akai-shuuichi** [comment]
我遗漏了一点，在多工具调用的情况下，工具的返回顺序会影响最终的正确结果，这不会受到tool_call_id的影响，但在qwen2.5的模版中并没有体现，如果希望在多个tool返回顺序和调用顺序不同情况下仍能保持精准，我认为可以在prompt中添加对应的tool_call_id, 但我认为这个功能应该由用户自行实现 :)

File: docs/en/llm/api_server_tools.md [hunk 7]
```diff
@@ -241,3 +241,130 @@ messages += [
...
+from openai import OpenAI
+import json
+
+def get_current_temperature(arguments):
+    if json.loads(arguments).get('location') == 'San Francisco, CA, USA':
+        return "26.0­°C"
+    else:
+        return "27.0"
+
+def get_temperature_date(arguments):
```
Comment (lvhan028) [inline docs/en/llm/api_server_tools.md:259]: 函数名的含义和函数体的实现感觉不匹配。这里是想表达获取某日期的气温么？

File: lmdeploy/model.py [hunk 8]
```diff
@@ -927,14 +927,122 @@ def match(cls, model_path: str) -> Optional[str]:
...
+            str: the concatenated prompt
+        """
+        if isinstance(messages, str):
+            return self.get_prompt(messages, sequence_start)
+        box_map = dict(user=self.user,
+                       assistant=self.assistant,
+                       system=self.system)
+        ret = ''
+        tool_prompt = ''
+        if tools is not None:
```
Comment (lvhan028) [inline lmdeploy/model.py:991]: 看起来拼接 messages 的逻辑还挺复杂的，最好在 tests/test_lmdeploy/test_model.py 中为这个函数增加单元测试用例，覆盖 if-else分支。有助于提前发现一些问题

File: docs/en/llm/api_server_tools.md (hunk 7 above)
Comment (akai-shuuichi) [reply docs/en/llm/api_server_tools.md:259]: 是的，不过我的实现有些粗糙了，参考了
https://qwen.readthedocs.io/zh-cn/latest/framework/function_call.html
我应当把函数签名写的更准确些
```
def get_temperature_date(location: str, date: str, unit: str = "celsius"):
```

File: lmdeploy/model.py (hunk 8 above)
Comment (akai-shuuichi) [reply lmdeploy/model.py:991]: 是的，我应当通过
```
tokenizer = AutoTokenizer.from_pretrained(model)
tokenizer.apply_chat_template(messages, tools=tools, add_generation_prompt=True, tokenize=False)
```
生成的结果进行比对

**Comment by akai-shuuichi** [comment]
我在tests/test_lmdeploy/test_model.py中添加了test_qwen2d5测试，主要测试了单工具和多工具的单轮/多轮的prompt拼接情况，预期prompt采用了apply_chat_template生成
```
tokenizer = AutoTokenizer.from_pretrained("/Users/zixianggao/Desktop/study/Qwen2.5-32B-Instruct-AWQ")
text = tokenizer.apply_chat_template(messages, tools=tools, add_generation_prompt=True, tokenize=False)
```

**Pull request merged**